File: 0234-palindrome-linked-list-回文链表/0234.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # O(n)时间复杂度, O(1)空间复杂度, 最后将链表结构恢复
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next: return True
        slow = head
        fast = head
        last = head
        nextNode = slow.next
        while slow:
            if fast.next and fast.next.next:
                fast = fast.next.next
                slow = nextNode
                nextNode = slow.next
                slow.next = last
                last = slow
            else:
                if not fast.next:   # 奇
                    fast = last.next
                    slow = nextNode
                    break
                else:               # 偶
                    fast = last
                    slow = nextNode
                    break

        # fast往前遍历, slow继续向后遍历
        # 恢复链表结构
        nextNode = fast.next
        while slow:
            if slow.val != fast.val: return False
            slow = slow.next
            fast = nextNode
            nextNode = fast.next
            fast.next = last
            last = fast

        return True

File: 0234-palindrome-linked-list-回文链表/0234.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if (!head || !head.next) return true;
    slow = head;
    fast = head;
    next = slow.next;
    while (fast.next && fast.next.next) {
        fast = fast.next.next;
        last = slow;
        slow = next;
        next = slow.next;
        slow.next = last;
    }

    // fast往回, slow往后
    if (fast.next) {        // 偶
        fast = slow;
        slow = next;
    } else {                // 奇
        fast = slow.next;
        slow = next;
    }

    next = fast.next;
    while (slow) {
        if (fast.val != slow.val) {
            return false
        }
        slow = slow.next
        last = fast;
        fast = next;
        next = fast.next;
        fast.next = last;
    }
    return true;
};